library/src/scala/Tuple.scala

```diff
@@ -82,82 +81,131 @@ sealed trait Tuple extends Product {
   /** Given a tuple `(a1, ..., am)`, returns the reversed tuple `(am, ..., a1)`
    *  consisting all its elements.
    */
-  @experimental
   inline def reverse[This >: this.type <: Tuple]: Reverse[This] =
     runtime.Tuples.reverse(this).asInstanceOf[Reverse[This]]
-}
 
-object Tuple {
+  /** A tuple with the elements of this tuple in reversed order added in front of `acc` */
+  inline def reverseOnto[This >: this.type <: Tuple, Acc <: Tuple](acc: Acc): ReverseOnto[This, Acc] =
+    (this.reverse ++ acc).asInstanceOf[ReverseOnto[This, Acc]]
 
-  /** Type of a tuple with an element appended */
-  type Append[X <: Tuple, Y] <: NonEmptyTuple = X match {
-    case EmptyTuple => Y *: EmptyTuple
-    case x *: xs => x *: Append[xs, Y]
-  }
+  /** A tuple consisting of all elements of this tuple that have types
+   *  for which the given type level predicate `P` reduces to the literal
+   *  constant `true`.
+   */
+  inline def filter[This >: this.type <: Tuple, P[_] <: Boolean]: Filter[This, P] =
```

Comment:
this method seems pretty irregular, i.e. why does it only operate on the type of the element?

and if it is intended to be used with literal types, then why not `constValueTuple[Filter[This, P]]`. I think we should review these additions before merging, unless we make them experimental.

Reply:
It would be quite the same thing, the elements types could be specific enough for the filter predicate `P` but not enough for `constValue`.
For example, the following could only compile with the current implementation of filter
```scala
type IsInt[X] = X match
  case Int => true
  case _ => false

def foo(tup: (Int, Int, String)) = tup.filter[T, IsInt]
```

Reply:
That being said, I agree the disconnect between having the filtering predicate on the type level Vs the querying of the filtered elements on the term level, could be a bit confusing.

Reply:
But for indexOf you two were arguing exactly the other way around? Make it type level, not value level?

Reply:
I believe every operation type on Tuple should have a corresponding term level operation. That was why filter was added.

Reply:
I think Jamie's point was that even with the two corresponding operations, the `type Filter` should have a type level predicate `P`, whereas the term level `def filter` should have a term level predicate. Similarly to what map does.